File: gdext/src/Viewports.cpp
#include "Viewports.h"
#include "Context.h"
#include "ImGuiController.h"
#include <imgui.h>

#include <godot_cpp/classes/display_server.hpp>
#include <godot_cpp/classes/rendering_server.hpp>
#include <godot_cpp/variant/callable.hpp>
#include <godot_cpp/variant/utility_functions.hpp>

namespace ImGui::Godot {

struct Godot_ViewportData
{
    Window* window = nullptr;
};

struct Viewports::Impl
{
    void InitPlatformInterface();
    void UpdateMonitors();
};

static void Godot_CreateWindow(ImGuiViewport* vp)
{
    Godot_ViewportData* vd = IM_NEW(Godot_ViewportData);
    vp->PlatformUserData = vd;

    {
        Godot_ViewportData* mainvd = (Godot_ViewportData*)ImGui::GetMainViewport()->PlatformUserData;
        Window* mainWindow = mainvd->window;
        if (mainWindow->is_embedding_subwindows())
        {
            if (ProjectSettings::get_singleton()->get("display/window/subwindows/embed_subwindows"))
            {
                UtilityFunctions::push_warning(
                    "ImGui Viewports: 'display/window/subwindows/embed_subwindows' needs to be disabled");
            }
            mainWindow->set_embedding_subwindows(false);
        }
    }

    Rect2i winRect = Rect2i(vp->Pos, vp->Size);

    ImGuiWindow* igwin = memnew(ImGuiWindow);
    igwin->init(vp);
    vd->window = Object::cast_to<Window>(igwin);
    vd->window->set_flag(Window::FLAG_BORDERLESS, true);
    vd->window->set_position(winRect.position);
    vd->window->set_size(winRect.size);
    vd->window->set_transparent_background(true);
    vd->window->set_flag(Window::FLAG_TRANSPARENT, true);
    vd->window->set_flag(Window::FLAG_ALWAYS_ON_TOP, vp->Flags & ImGuiViewportFlags_TopMost);

    Node* root = Object::cast_to<Node>(Engine::get_singleton()->get_singleton("ImGuiController"));
    root->add_child(vd->window);

    // need to do this after add_child
    vd->window->set_flag(Window::FLAG_TRANSPARENT, true);

    // it's our window, so just draw directly to the root viewport
    RID vprid = vd->window->get_viewport_rid();
    vp->RendererUserData = (void*)vprid.get_id();

    int32_t windowID = vd->window->get_window_id();
    DisplayServer::get_singleton()->window_set_input_event_callback(
        Callable(ImGuiController::Instance(), "window_input_callback"),
        windowID);

    RenderingServer* RS = RenderingServer::get_singleton();
    ImGui::Godot::GetContext()->renderer->InitViewport(vprid);
    RS->viewport_set_transparent_background(vprid, true);
}

static void Godot_DestroyWindow(ImGuiViewport* vp)
{
    Godot_ViewportData* vd = (Godot_ViewportData*)vp->PlatformUserData;
    if (vd)
    {
        if (vd->window && vp != ImGui::GetMainViewport())
        {
            vd->window->get_parent()->remove_child(vd->window);
            memdelete(vd->window);
        }
        IM_DELETE(vd);
        vp->PlatformUserData = nullptr;
        vp->RendererUserData = nullptr;
    }
}

static void Godot_ShowWindow(ImGuiViewport* vp)
{
    Godot_ViewportData* vd = (Godot_ViewportData*)vp->PlatformUserData;
    vd->window->show();
}

static void Godot_SetWindowPos(ImGuiViewport* vp, ImVec2 pos)
{
    Godot_ViewportData* vd = (Godot_ViewportData*)vp->PlatformUserData;
    vd->window->set_position(pos);
}

ImVec2 Godot_GetWindowPos(ImGuiViewport* vp)
{
    Godot_ViewportData* vd = (Godot_ViewportData*)vp->PlatformUserData;
    return vd->window->get_position();
}

void Godot_SetWindowSize(ImGuiViewport* vp, ImVec2 size)
{
    Godot_ViewportData* vd = (Godot_ViewportData*)vp->PlatformUserData;
    vd->window->set_size(size);
}

ImVec2 Godot_GetWindowSize(ImGuiViewport* vp)
{
    Godot_ViewportData* vd = (Godot_ViewportData*)vp->PlatformUserData;
    return vd->window->get_size();
}

void Godot_SetWindowFocus(ImGuiViewport* vp)
{
    Godot_ViewportData* vd = (Godot_ViewportData*)vp->PlatformUserData;
    vd->window->grab_focus();
}

bool Godot_GetWindowFocus(ImGuiViewport* vp)
{
    Godot_ViewportData* vd = (Godot_ViewportData*)vp->PlatformUserData;
    return vd->window->has_focus();
}

bool Godot_GetWindowMinimized(ImGuiViewport* vp)
{
    Godot_ViewportData* vd = (Godot_ViewportData*)vp->PlatformUserData;
    return vd->window->get_mode() & Window::MODE_MINIMIZED;
}

void Godot_SetWindowTitle(ImGuiViewport* vp, const char* str)
{
    Godot_ViewportData* vd = (Godot_ViewportData*)vp->PlatformUserData;
    vd->window->set_title(String(str));
}

void Viewports::Impl::InitPlatformInterface()
{
    auto& pio = ImGui::GetPlatformIO();
    pio.Platform_CreateWindow = Godot_CreateWindow;
    pio.Platform_DestroyWindow = Godot_DestroyWindow;
    pio.Platform_ShowWindow = Godot_ShowWindow;
    pio.Platform_SetWindowPos = Godot_SetWindowPos;
    pio.Platform_GetWindowPos = Godot_GetWindowPos;
    pio.Platform_SetWindowSize = Godot_SetWindowSize;
    pio.Platform_GetWindowSize = Godot_GetWindowSize;
    pio.Platform_SetWindowFocus = Godot_SetWindowFocus;
    pio.Platform_GetWindowFocus = Godot_GetWindowFocus;
    pio.Platform_GetWindowMinimized = Godot_GetWindowMinimized;
    pio.Platform_SetWindowTitle = Godot_SetWindowTitle;
}

void Viewports::Impl::UpdateMonitors()
{
    auto& pio = ImGui::GetPlatformIO();
    DisplayServer* DS = DisplayServer::get_singleton();
    int screenCount = DS->get_screen_count();

    pio.Monitors.resize(0);
    for (int i = 0; i < screenCount; ++i)
    {
        ImGuiPlatformMonitor monitor;
        monitor.MainPos = DS->screen_get_position(i);
        monitor.MainSize = DS->screen_get_size(i);
        monitor.DpiScale = DS->screen_get_scale(i);

        Rect2i rect = DS->screen_get_usable_rect(i);
        monitor.WorkPos = rect.position;
        monitor.WorkSize = rect.size;

        pio.Monitors.push_back(monitor);
    }

    // fix headless
    if (pio.Monitors.size() == 0)
    {
        ImGuiPlatformMonitor monitor;
        monitor.MainPos = {0, 0};
        monitor.MainSize = {640, 480};
        monitor.DpiScale = 1.0f;

        monitor.WorkPos = {0, 0};
        monitor.WorkSize = {640, 480};

        pio.Monitors.push_back(monitor);
    }
}

Viewports::Viewports() : impl(std::make_unique<Impl>())
{
    auto& io = ImGui::GetIO();
    io.BackendFlags |= ImGuiBackendFlags_PlatformHasViewports;
    impl->InitPlatformInterface();
    impl->UpdateMonitors();
}

void Viewports::SetMainWindow(Window* mainWindow, RID mainSubViewport)
{
    Godot_ViewportData* mainvd = IM_NEW(Godot_ViewportData);
    mainvd->window = mainWindow;
    ImGuiViewport* vp = ImGui::GetMainViewport();
    vp->PlatformUserData = mainvd;
    vp->RendererUserData = (void*)mainSubViewport.get_id();
}

Viewports::~Viewports()
{
}

} // namespace ImGui::Godot

File: addons/imgui-godot/ImGuiGodot/Internal/Viewports.cs
#if GODOT_PC
using Godot;
using Hexa.NET.ImGui;
using System;
using System.Runtime.InteropServices;
using Vector2 = System.Numerics.Vector2;

namespace ImGuiGodot.Internal;

internal sealed class GodotImGuiWindow : IDisposable
{
    private readonly GCHandle _gcHandle;
    private readonly ImGuiViewportPtr _vp;
    private readonly Window _window;
    private readonly bool _isOwnedWindow = false;

    /// <summary>
    /// sub window
    /// </summary>
    public unsafe GodotImGuiWindow(ImGuiViewportPtr vp)
    {
        _gcHandle = GCHandle.Alloc(this);
        _vp = vp;
        _vp.PlatformUserData = (void*)GCHandle.ToIntPtr(_gcHandle);
        _isOwnedWindow = true;

        Rect2I winRect = new(_vp.Pos.ToVector2I(), _vp.Size.ToVector2I());

        Window mainWindow = ImGuiController.Instance.GetWindow();
        if (mainWindow.GuiEmbedSubwindows)
        {
            if ((bool)ProjectSettings.GetSetting("display/window/subwindows/embed_subwindows"))
            {
                GD.PushWarning(
                    "ImGui Viewports: 'display/window/subwindows/embed_subwindows' needs to be disabled");
            }
            mainWindow.GuiEmbedSubwindows = false;
        }

        _window = new Window()
        {
            Borderless = true,
            Position = winRect.Position,
            Size = winRect.Size,
            Transparent = true,
            TransparentBg = true,
            AlwaysOnTop = vp.Flags.HasFlag(ImGuiViewportFlags.TopMost),
            Unfocusable = vp.Flags.HasFlag(ImGuiViewportFlags.NoFocusOnClick)
        };

        _window.CloseRequested += () => _vp.PlatformRequestClose = true;
        _window.SizeChanged += () => _vp.PlatformRequestResize = true;
        _window.WindowInput += ImGuiController.WindowInputCallback;

        ImGuiController.Instance.AddChild(_window);

        // need to do this after AddChild
        _window.Transparent = true;

        // it's our window, so just draw directly to the root viewport
        var vprid = _window.GetViewportRid();
        _vp.RendererUserData = (void*)vprid.Id;
        _vp.PlatformHandle = (void*)_window.GetWindowId();

        State.Instance.Renderer.InitViewport(vprid);
        RenderingServer.ViewportSetTransparentBackground(_window.GetViewportRid(), true);
    }

    /// <summary>
    /// main window
    /// </summary>
    public unsafe GodotImGuiWindow(ImGuiViewportPtr vp, Window gw, Rid mainSubViewport)
    {
        _gcHandle = GCHandle.Alloc(this);
        _vp = vp;
        _vp.PlatformUserData = (void*)GCHandle.ToIntPtr(_gcHandle);
        _window = gw;
        _vp.RendererUserData = (void*)mainSubViewport.Id;
    }

    public unsafe void Dispose()
    {
        if (_gcHandle.IsAllocated)
        {
            if (_isOwnedWindow)
            {
                State.Instance.Renderer
                    .CloseViewport(Util.ConstructRid((ulong)_vp.RendererUserData));
                _window.GetParent().RemoveChild(_window);
                _window.Free();
            }
            _gcHandle.Free();
        }
    }

    public void ShowWindow()
    {
        _window.Show();
    }

    public void SetWindowPos(Vector2I pos)
    {
        _window.Position = pos;
    }

    public Vector2I GetWindowPos()
    {
        return _window.Position;
    }

    public void SetWindowSize(Vector2I size)
    {
        _window.Size = size;
    }

    public Vector2I GetWindowSize()
    {
        return _window.Size;
    }

    public void SetWindowFocus()
    {
        _window.GrabFocus();
    }

    public bool GetWindowFocus()
    {
        return _window.HasFocus();
    }

    public bool GetWindowMinimized()
    {
        return _window.Mode.HasFlag(Window.ModeEnum.Minimized);
    }

    public void SetWindowTitle(string title)
    {
        _window.Title = title;
    }
}

internal static class ViewportsExts
{
    internal static Vector2 ToImVec2(this Vector2I v)
    {
        return new Vector2(v.X, v.Y);
    }

    internal static Vector2I ToVector2I(this Vector2 v)
    {
        return new Vector2I((int)v.X, (int)v.Y);
    }
}

internal sealed partial class Viewports
{
    [LibraryImport("cimgui")]
    [UnmanagedCallConv(CallConvs = [typeof(System.Runtime.CompilerServices.CallConvCdecl)])]
    private static unsafe partial void ImGuiPlatformIO_Set_Platform_GetWindowPos(
        ImGuiPlatformIO* platform_io,
        IntPtr funcPtr);
    [LibraryImport("cimgui")]
    [UnmanagedCallConv(CallConvs = [typeof(System.Runtime.CompilerServices.CallConvCdecl)])]
    private static unsafe partial void ImGuiPlatformIO_Set_Platform_GetWindowSize(
        ImGuiPlatformIO* platform_io,
        IntPtr funcPtr);

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate void Platform_CreateWindow(ImGuiViewportPtr vp);
    private static readonly Platform_CreateWindow _createWindow = Godot_CreateWindow;

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate void Platform_DestroyWindow(ImGuiViewportPtr vp);
    private static readonly Platform_DestroyWindow _destroyWindow = Godot_DestroyWindow;

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate void Platform_ShowWindow(ImGuiViewportPtr vp);
    private static readonly Platform_ShowWindow _showWindow = Godot_ShowWindow;

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate void Platform_SetWindowPos(ImGuiViewportPtr vp, Vector2 pos);
    private static readonly Platform_SetWindowPos _setWindowPos = Godot_SetWindowPos;

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate void Platform_GetWindowPos(ImGuiViewportPtr vp, out Vector2 pos);
    private static readonly Platform_GetWindowPos _getWindowPos = Godot_GetWindowPos;

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate void Platform_SetWindowSize(ImGuiViewportPtr vp, Vector2 pos);
    private static readonly Platform_SetWindowSize _setWindowSize = Godot_SetWindowSize;

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate void Platform_GetWindowSize(ImGuiViewportPtr vp, out Vector2 size);
    private static readonly Platform_GetWindowSize _getWindowSize = Godot_GetWindowSize;

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate void Platform_SetWindowFocus(ImGuiViewportPtr vp);
    private static readonly Platform_SetWindowFocus _setWindowFocus = Godot_SetWindowFocus;

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate bool Platform_GetWindowFocus(ImGuiViewportPtr vp);
    private static readonly Platform_GetWindowFocus _getWindowFocus = Godot_GetWindowFocus;

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate bool Platform_GetWindowMinimized(ImGuiViewportPtr vp);
    private static readonly Platform_GetWindowMinimized _getWindowMinimized
        = Godot_GetWindowMinimized;

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate void Platform_SetWindowTitle(ImGuiViewportPtr vp, string title);
    private static readonly Platform_SetWindowTitle _setWindowTitle = Godot_SetWindowTitle;

    private GodotImGuiWindow _mainWindow = null!;

    private static void UpdateMonitors()
    {
        var pio = ImGui.GetPlatformIO();
        int screenCount = DisplayServer.GetScreenCount();

        // workaround for lack of ImVector constructor

        pio.Monitors.Clear();
        pio.Monitors.Resize(screenCount);

        for (int i = 0; i < screenCount; ++i)
        {
            var rect = DisplayServer.ScreenGetUsableRect(i);
            pio.Monitors[i] = new()
            {
                MainPos = DisplayServer.ScreenGetPosition(i).ToImVec2(),
                MainSize = DisplayServer.ScreenGetSize(i).ToImVec2(),
                DpiScale = DisplayServer.ScreenGetScale(i),
                WorkPos = rect.Position.ToImVec2(),
                WorkSize = rect.Size.ToImVec2()
            };
        }

        // TODO: add monitor if headless
    }

    private static unsafe void InitPlatformInterface()
    {
        var pio = ImGui.GetPlatformIO();

        pio.PlatformCreateWindow = (void*)Marshal.GetFunctionPointerForDelegate(_createWindow);
        pio.PlatformDestroyWindow = (void*)Marshal.GetFunctionPointerForDelegate(_destroyWindow);
        pio.PlatformShowWindow = (void*)Marshal.GetFunctionPointerForDelegate(_showWindow);
        pio.PlatformSetWindowPos = (void*)Marshal.GetFunctionPointerForDelegate(_setWindowPos);
        //pio->Platform_GetWindowPos = Marshal.GetFunctionPointerForDelegate(_getWindowPos);
        pio.PlatformSetWindowSize = (void*)Marshal.GetFunctionPointerForDelegate(_setWindowSize);
        //pio->Platform_GetWindowSize = Marshal.GetFunctionPointerForDelegate(_getWindowSize);
        pio.PlatformSetWindowFocus = (void*)Marshal.GetFunctionPointerForDelegate(_setWindowFocus);
        pio.PlatformGetWindowFocus = (void*)Marshal.GetFunctionPointerForDelegate(_getWindowFocus);
        pio.PlatformGetWindowMinimized = (void*)Marshal.GetFunctionPointerForDelegate(
            _getWindowMinimized);
        pio.PlatformSetWindowTitle = (void*)Marshal.GetFunctionPointerForDelegate(_setWindowTitle);

        ImGuiPlatformIO_Set_Platform_GetWindowPos(
            pio,
            Marshal.GetFunctionPointerForDelegate(_getWindowPos));
        ImGuiPlatformIO_Set_Platform_GetWindowSize(
            pio,
            Marshal.GetFunctionPointerForDelegate(_getWindowSize));
    }

    public Viewports()
    {
        InitPlatformInterface();
        UpdateMonitors();
    }

    public void SetMainWindow(Window window, Rid mainSubViewport)
    {
        _mainWindow?.Dispose();
        _mainWindow = new GodotImGuiWindow(ImGui.GetMainViewport(), window, mainSubViewport);
    }

    private static void Godot_CreateWindow(ImGuiViewportPtr vp)
    {
        _ = new GodotImGuiWindow(vp);
    }

    private static unsafe void Godot_DestroyWindow(ImGuiViewportPtr vp)
    {
        if (vp.PlatformUserData != null)
        {
            var window = (GodotImGuiWindow)GCHandle.FromIntPtr((nint)vp.PlatformUserData).Target!;
            window.Dispose();
            vp.PlatformUserData = null;
            vp.RendererUserData = null;
        }
    }

    private static unsafe void Godot_ShowWindow(ImGuiViewportPtr vp)
    {
        var window = (GodotImGuiWindow)GCHandle.FromIntPtr((nint)vp.PlatformUserData).Target!;
        window.ShowWindow();
    }

    private static unsafe void Godot_SetWindowPos(ImGuiViewportPtr vp, Vector2 pos)
    {
        var window = (GodotImGuiWindow)GCHandle.FromIntPtr((nint)vp.PlatformUserData).Target!;
        window.SetWindowPos(pos.ToVector2I());
    }

    private static unsafe void Godot_GetWindowPos(ImGuiViewportPtr vp, out Vector2 pos)
    {
        var window = (GodotImGuiWindow)GCHandle.FromIntPtr((nint)vp.PlatformUserData).Target!;
        pos = window.GetWindowPos().ToImVec2();
    }

    private static unsafe void Godot_SetWindowSize(ImGuiViewportPtr vp, Vector2 size)
    {
        var window = (GodotImGuiWindow)GCHandle.FromIntPtr((nint)vp.PlatformUserData).Target!;
        window.SetWindowSize(size.ToVector2I());
    }

    private static unsafe void Godot_GetWindowSize(ImGuiViewportPtr vp, out Vector2 size)
    {
        var window = (GodotImGuiWindow)GCHandle.FromIntPtr((nint)vp.PlatformUserData).Target!;
        size = window.GetWindowSize().ToImVec2();
    }

    private static unsafe void Godot_SetWindowFocus(ImGuiViewportPtr vp)
    {
        var window = (GodotImGuiWindow)GCHandle.FromIntPtr((nint)vp.PlatformUserData).Target!;
        window.SetWindowFocus();
    }

    private static unsafe bool Godot_GetWindowFocus(ImGuiViewportPtr vp)
    {
        var window = (GodotImGuiWindow)GCHandle.FromIntPtr((nint)vp.PlatformUserData).Target!;
        return window.GetWindowFocus();
    }

    private static unsafe bool Godot_GetWindowMinimized(ImGuiViewportPtr vp)
    {
        var window = (GodotImGuiWindow)GCHandle.FromIntPtr((nint)vp.PlatformUserData).Target!;
        return window.GetWindowMinimized();
    }

    private static unsafe void Godot_SetWindowTitle(ImGuiViewportPtr vp, string title)
    {
        var window = (GodotImGuiWindow)GCHandle.FromIntPtr((nint)vp.PlatformUserData).Target!;
        window.SetWindowTitle(title);
    }
}
#endif
